plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSourceProvider.java

```diff
@@ -307,46 +314,62 @@ public ConnectorPageSource createPageSource(
 
         Optional<PartitionData> partition = partitionSpec.isUnpartitioned() ? Optional.empty() : Optional.of(partitionData);
         LocationProvider locationProvider = getLocationProvider(table.getSchemaTableName(), table.getTableLocation(), table.getStorageProperties());
-        Supplier<IcebergPositionDeletePageSink> positionDeleteSink = () -> new IcebergPositionDeletePageSink(
-                split.getPath(),
-                partitionSpec,
-                partition,
-                locationProvider,
-                fileWriterFactory,
-                hdfsEnvironment,
-                hdfsContext,
-                jsonCodec,
-                session,
-                split.getFileFormat(),
-                table.getStorageProperties(),
-                split.getFileRecordCount());
-
-        Supplier<IcebergPageSink> updatedRowPageSinkSupplier = () -> new IcebergPageSink(
-                tableSchema,
-                PartitionSpecParser.fromJson(tableSchema, table.getPartitionSpecJson()),
-                locationProvider,
-                fileWriterFactory,
-                pageIndexerFactory,
-                hdfsEnvironment,
-                hdfsContext,
-                tableSchema.columns().stream().map(column -> getColumnHandle(column, typeManager)).collect(toImmutableList()),
-                jsonCodec,
-                session,
-                split.getFileFormat(),
-                table.getStorageProperties(),
-                maxOpenPartitions);
+        Supplier<ConnectorPageSink> positionDeleteSink = () -> {
+            try (ThreadContextClassLoader ignored = new ThreadContextClassLoader(classLoader)) {
+                // Make sure we are using the PluginClassLoader when initialising and using the sink instance
+                return new ClassLoaderSafeConnectorPageSink(
+                        new IcebergPositionDeletePageSink(
+                                split.getPath(),
+                                partitionSpec,
+                                partition,
+                                locationProvider,
+                                fileWriterFactory,
+                                hdfsEnvironment,
+                                hdfsContext,
+                                jsonCodec,
+                                session,
+                                split.getFileFormat(),
+                                table.getStorageProperties(),
+                                split.getFileRecordCount()),
+                        classLoader);
+            }
+        };
+
+        Supplier<ConnectorPageSink> updatedRowPageSinkSupplier = () -> {
+            try (ThreadContextClassLoader ignored = new ThreadContextClassLoader(classLoader)) {
+                // Make sure we are using the PluginClassLoader when initialising and using the sink instance
+                return new ClassLoaderSafeConnectorPageSink(
```

Comment:
Same here, maybe I am missing something but I think it is enough to use `ClassLoaderSafeConnectorPageSink` without try/catch above. You can see for example how `ClassLoaderSafeConnectorPageSinkProvider` is used

Reply:
Also I don't understand why here you did this double wrapping and below with `IcebergPageSource` you didn't

Reply:
https://github.com/trinodb/trino/pull/13078#discussion_r913230518